tests/integration_tests/functional/test_snapshot_restore_cross_kernel.py

```diff
@@ -0,0 +1,178 @@
+# Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Test to restore snapshots across kernel versions."""
+import json
+import logging
+import os
+import pathlib
+import shutil
+import pytest
+
+from framework.artifacts import Snapshot, Artifact, ArtifactType, \
+    create_net_devices_configuration
+from framework.builder import MicrovmBuilder
+from framework.defs import FC_WORKSPACE_DIR, DEFAULT_TEST_SESSION_ROOT_PATH
+from framework.utils_vsock import check_vsock_device
+from integration_tests.functional.test_mmds import _populate_data_store
+from integration_tests.functional.test_snapshot_basic import \
+    _guest_run_fio_iteration
+from integration_tests.functional.test_balloon import \
+    get_stable_rss_mem_by_pid, make_guest_dirty_memory, MB_TO_PAGES
+import host_tools.network as net_tools  # pylint: disable=import-error
+
+
+# Define 4 net device configurations.
+net_ifaces = create_net_devices_configuration(4)
+
+
+def _test_balloon(microvm, ssh_connection):
+    # Get the firecracker pid.
+    firecracker_pid = microvm.jailer_clone_pid
+
+    # Check memory usage.
+    first_reading = get_stable_rss_mem_by_pid(firecracker_pid)
+    # Dirty 60MB of pages.
+    make_guest_dirty_memory(ssh_connection, amount=(60 * MB_TO_PAGES))
+    # Check memory usage again.
+    second_reading = get_stable_rss_mem_by_pid(firecracker_pid)
+    assert second_reading > first_reading
+
+    # Inflate the balloon.
+    response = microvm.balloon.patch(amount_mib=40)
+    assert microvm.api_session.is_status_no_content(response.status_code)
+
+    third_reading = get_stable_rss_mem_by_pid(firecracker_pid)
+    # Ensure that there is a reduction in RSS.
+    assert second_reading > third_reading
+
+
+def _get_snapshot_files_paths(snapshot_dir):
+    mem = vmstate = ssh_key = disk = None
+
+    for file in os.listdir(snapshot_dir):
+        file_path = os.path.join(Artifact.LOCAL_ARTIFACT_DIR, file)
+
+        if file.endswith(".mem"):
+            mem = file_path
+        elif file.endswith(".vmstate"):
+            vmstate = file_path
+        elif file.endswith(".id_rsa"):
+            ssh_key = Artifact(None, os.path.basename(file),
+                               ArtifactType.SSH_KEY,
+                               DEFAULT_TEST_SESSION_ROOT_PATH)
+            file_path = ssh_key.local_path()
+            pathlib.Path(os.path.dirname(file_path)).mkdir(parents=True,
```

Comment:
why is this mkdir needed?

Reply:
This is needed because `build_from_snapshot` expects an Artifact object to be passed to `ssh_key` inside the `Snapshot` object. When making use of that ssh key, the `local_path()` function returns a path that includes the artifact type and also the platform (something like: ` <path>/<artifact_type>/<platform>/<ssh_key_name>`), so here we create that dir path.